File: jserver-cypress/src/main/java/com/github/gr3gdev/jserver/test/CypressTest.java
package com.github.gr3gdev.jserver.test;

import com.github.gr3gdev.jserver.annotations.Plugin;
import com.github.gr3gdev.jserver.security.TokenClientPlugin;
import com.github.gr3gdev.jserver.security.TokenServerPlugin;
import com.github.gr3gdev.jserver.thymeleaf.ThymeleafPlugin;
import org.bouncycastle.asn1.x509.ExtendedKeyUsage;
import org.bouncycastle.asn1.x509.KeyPurposeId;
import org.bouncycastle.asn1.x509.X509Extensions;
import org.bouncycastle.asn1.x509.X509Name;
import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.bouncycastle.x509.X509V3CertificateGenerator;

import javax.security.auth.x500.X500Principal;
import java.io.FileOutputStream;
import java.math.BigInteger;
import java.security.*;
import java.security.cert.X509Certificate;
import java.util.Date;

public class CypressTest {

    private final PublicKey publicKey;
    private final PrivateKey privateKey;

    public CypressTest() throws Exception {
        Security.addProvider(new BouncyCastleProvider());
        final String keystorePassword = "MyDemoP@ssw0RD";

        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
        ks.load(null, keystorePassword.toCharArray());

        KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA");
        keyPairGenerator.initialize(2048);
        KeyPair keyPair = keyPairGenerator.generateKeyPair();

        publicKey = keyPair.getPublic();
        privateKey = keyPair.getPrivate();

        X509Certificate[] certificateChain = new X509Certificate[1];
        certificateChain[0] = generateSelfSignedCertificate(keyPair);
        ks.setKeyEntry("test", keyPair.getPrivate(), keystorePassword.toCharArray(), certificateChain);

        try (FileOutputStream fos = new FileOutputStream("/tmp/demo.jks")) {
            ks.store(fos, keystorePassword.toCharArray());
        }

    }

    private static X509Certificate generateSelfSignedCertificate(KeyPair keyPair) throws Exception {
        X509V3CertificateGenerator certGen = new X509V3CertificateGenerator();
        X500Principal dnName = new X500Principal("cn=example");
        certGen.setSerialNumber(BigInteger.valueOf(System.currentTimeMillis()));
        certGen.setSubjectDN(new X509Name("dc=name"));
        certGen.setIssuerDN(dnName);
        certGen.setNotBefore(new Date(System.currentTimeMillis() - 24 * 60 * 60 * 1000));
        certGen.setNotAfter(new Date(System.currentTimeMillis() + 2L * 365 * 24 * 60 * 60 * 1000));
        certGen.setPublicKey(keyPair.getPublic());
        certGen.setSignatureAlgorithm("SHA256WithRSAEncryption");
        certGen.addExtension(X509Extensions.ExtendedKeyUsage, true,
                new ExtendedKeyUsage(KeyPurposeId.id_kp_timeStamping));
        return certGen.generate(keyPair.getPrivate(), "BC");
    }

    @Plugin
    public TokenServerPlugin tokenServerPlugin() {
        return new TokenServerPlugin(privateKey)
                .id("jServer-cypress")
                .issuer("GR3Gdev");
    }

    @Plugin
    public TokenClientPlugin tokenClientPlugin() {
        return new TokenClientPlugin(publicKey);
    }

    @Plugin
    public ThymeleafPlugin thymeleafPlugin() {
        return new ThymeleafPlugin();
    }

}

File: jserver-cypress/src/main/kotlin/com/github/gr3gdev/jserver/test/CypressTest.kt
package com.github.gr3gdev.jserver.test

import com.github.gr3gdev.jserver.JServer
import com.github.gr3gdev.jserver.logger.Logger
import com.github.gr3gdev.jserver.route.HttpStatus
import com.github.gr3gdev.jserver.route.Response
import com.github.gr3gdev.jserver.route.RouteListener
import com.github.gr3gdev.jserver.security.TokenClientPlugin
import com.github.gr3gdev.jserver.security.TokenServerPlugin
import com.github.gr3gdev.jserver.test.route.*
import com.github.gr3gdev.jserver.thymeleaf.ThymeleafPlugin
import java.security.KeyFactory
import java.security.spec.X509EncodedKeySpec
import java.util.*

class CypressTest

fun main() {
    Logger.changeLevel(Logger.Level.DEBUG)
    val publicFile = String(SecureRoute::class.java.getResourceAsStream("/public.txt").readAllBytes())
            .replace("\n", "")
            .replace("-----BEGIN PUBLIC KEY-----", "")
            .replace("-----END PUBLIC KEY-----", "")
    val spec = X509EncodedKeySpec(Base64.getDecoder().decode(publicFile))

    val server = JServer.server()
            .plugins(
                    TokenServerPlugin("demo", CypressTest::class.java.getResourceAsStream("/demo.jks"), "MyDemoP@ssw0RD".toCharArray())
                            .id("jServerDemo").issuer("GR3Gdev"),
                    TokenClientPlugin(KeyFactory.getInstance("RSA").generatePublic(spec)),
                    ThymeleafPlugin()
            )
            // REACT front
            .get("/react/{file}", ReactRoute.static("/react"))
            .get("/react", ReactRoute.get())
            // API REST
            .get("/api/persons/{id}", ApiRoute.findById())
            .get("/api/persons", ApiRoute.findAll())
            .put("/api/persons", ApiRoute.save())
            // SECURE
            .get("/secure", SecureRoute.get())
            .get("/login", LoginRoute.get())
            .post("/login", LoginRoute.post())
            // THYMELEAF
            .get("/page1", ThymeleafRoute.get())
            .get("/css/{file}", RouteListener(HttpStatus.OK, Response.File("/css/test.css", "text/css")))
    // STOP
    server.get("/stop", StopRoute.exec(server))
            .start()
}
